vortexor/src/load_balancer.rs

```diff
@@ -0,0 +1,206 @@
+                    commitment: CommitmentLevel::Processed,
+                },
+            ));
+            if let Err(e) = rpc_client.get_slot() {
+                error!("error warming up rpc: {rpc_url}. error: {e}");
+            }
+            // store as ws instead of http so we can lookup by furthest ahead ws subscription
+            (ws.clone(), rpc_client)
+        }));
+
+        // sender tracked as health_manager-channel_stats.slot_sender_len
+        let (slot_sender, slot_receiver) = crossbeam_channel::bounded(Self::SLOT_QUEUE_CAPACITY);
+        let subscription_threads =
+            Self::start_subscription_threads(servers, server_to_slot.clone(), slot_sender, exit);
+        (
+            RpcLoadBalancer {
+                server_to_slot,
+                server_to_rpc_client,
+                subscription_threads,
+            },
+            slot_receiver,
+        )
+    }
+
+    /// Start threads to watch for slot events on the RPC servers and update
+    /// the server_to_slot map.
+    fn start_subscription_threads(
+        servers: &[(String, String)],
+        server_to_slot: Arc<DashMap<String, Slot>>,
+        slot_sender: Sender<Slot>,
+        exit: &Arc<AtomicBool>,
+    ) -> Vec<JoinHandle<()>> {
+        info!("start_subscription_threads for servers: {servers:?}");
+        let highest_slot = Arc::new(AtomicU64::default());
+
+        servers
+            .iter()
+            .map(|(_, websocket_url)| {
+                let ws_url_no_token = websocket_url
+                    .split('/')
+                    .nth(2)
+                    .unwrap_or_default()
+                    .to_string();
+                let exit = exit.clone();
+                let websocket_url = websocket_url.clone();
+                let server_to_slot = server_to_slot.clone();
+                let slot_sender = slot_sender.clone();
+                let highest_slot = highest_slot.clone();
+
+                Builder::new()
+                    .name(format!("load_balancer_subscription_thread-{ws_url_no_token}"))
+                    .spawn(move || {
+                        while !exit.load(Ordering::Relaxed) {
+                            info!("running slot_subscribe() with url: {websocket_url}");
+                            let mut last_slot_update = Instant::now();
+
+                            match PubsubClient::slot_subscribe(&websocket_url) {
+                                Ok((_subscription, receiver)) => {
+                                    while !exit.load(Ordering::Relaxed) {
+                                        match receiver.recv_timeout(Duration::from_millis(100))
+                                        {
+                                            Ok(slot) => {
+                                                last_slot_update = Instant::now();
+
+                                                server_to_slot
+                                                    .insert(websocket_url.clone(), slot.slot);
```

Comment:
insert the url initially (maybe with slot of 0?) then just mutate the value. we don't need to clone the string everytime